File: scripts/generate-random-state-variations.mjs
#!/usr/bin/env node
// Generate two perf-test variations of dashboard/data/All.json with random
// node states, designed to exercise every auto-collapse rule.
//
//   All-randomState.json              — toggleCollapseOnStatusChange: false
//   All-randomState-autoCollapse.json — toggleCollapseOnStatusChange: true
//
// State assignment is seeded so output is reproducible. Distribution and
// targeted subtree patches guarantee coverage of:
//   - Rule 1 (all children share one collapsible status: READY/UPDATED/SKIPPED/DISABLED)
//   - Rule 2 (children are only SKIPPED and/or UPDATED)
//   - Mixed-problem containers (ERROR/WARNING/DELAYED) that must stay expanded

import fs from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DATA_DIR = path.resolve(__dirname, '..', 'dashboard', 'data');
const INPUT = path.join(DATA_DIR, 'All.json');
const OUT_NO_COLLAPSE = path.join(DATA_DIR, 'All-randomState.json');
const OUT_COLLAPSE = path.join(DATA_DIR, 'All-randomState-autoCollapse.json');

// Status values match NodeStatus in dashboard/js/nodeBase.js
const STATES = {
  UNDETERMINED: 'Undetermined',
  UNKNOWN: 'Unknown',
  DISABLED: 'Disabled',
  READY: 'Ready',
  UPDATING: 'Updating',
  UPDATED: 'Updated',
  SKIPPED: 'Skipped',
  DELAYED: 'Delayed',
  WARNING: 'Warning',
  ERROR: 'Error',
};

// Mulberry32 — small deterministic PRNG so the variations are reproducible.
function mulberry32(seed) {
  let s = seed >>> 0;
  return () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

const rand = mulberry32(0xfd05_2026);

// Weighted distribution biased toward "success" so naturally-clustered small
// containers can fall under Rule 1 (all-same) while still producing enough
// problem states for mixed containers.
const WEIGHTED = [
  [STATES.UPDATED, 24],
  [STATES.READY, 18],
  [STATES.SKIPPED, 14],
  [STATES.UPDATING, 10],
  [STATES.DELAYED, 8],
  [STATES.WARNING, 8],
  [STATES.ERROR, 7],
  [STATES.DISABLED, 5],
  [STATES.UNKNOWN, 3],
  [STATES.UNDETERMINED, 3],
];
const totalWeight = WEIGHTED.reduce((a, [, w]) => a + w, 0);

function weightedPick() {
  let r = rand() * totalWeight;
  for (const [s, w] of WEIGHTED) {
    if ((r -= w) < 0) return s;
  }
  return STATES.UNKNOWN;
}

function pickFrom(arr) {
  return arr[Math.floor(rand() * arr.length)];
}

function* walkContainers(node, depth = 0) {
  if (Array.isArray(node.children) && node.children.length) {
    yield { node, depth };
    for (const c of node.children) yield* walkContainers(c, depth + 1);
  }
}

function* walkLeaves(node) {
  if (Array.isArray(node.children) && node.children.length) {
    for (const c of node.children) yield* walkLeaves(c);
  } else {
    yield node;
  }
}

function assignRandomStates(root) {
  const containers = [];
  for (const { node, depth } of walkContainers(root)) {
    containers.push({ node, depth });
  }

  // Default: every leaf gets a weighted-random state.
  for (const leaf of walkLeaves(root)) {
    leaf.state = weightedPick();
  }

  // Pick small leaf-only containers (parents whose children are all leaves)
  // and force patterns that exercise each auto-collapse rule.
  const leafOnlyContainers = containers.filter(
    ({ node }) =>
      node.children.length > 0 && node.children.every((c) => !c.children || !c.children.length),
  );

  // Shuffle deterministically.
  for (let i = leafOnlyContainers.length - 1; i > 0; i--) {
    const j = Math.floor(rand() * (i + 1));
    [leafOnlyContainers[i], leafOnlyContainers[j]] = [leafOnlyContainers[j], leafOnlyContainers[i]];
  }

  const patterns = [
    // Rule 1: all-same collapsible status
    (kids) => kids.forEach((k) => (k.state = STATES.READY)),
    (kids) => kids.forEach((k) => (k.state = STATES.UPDATED)),
    (kids) => kids.forEach((k) => (k.state = STATES.SKIPPED)),
    (kids) => kids.forEach((k) => (k.state = STATES.DISABLED)),
    // Rule 1: all-same non-collapsible status (must NOT collapse)
    (kids) => kids.forEach((k) => (k.state = STATES.ERROR)),
    (kids) => kids.forEach((k) => (k.state = STATES.WARNING)),
    // Rule 2: SKIPPED + UPDATED only mix
    (kids) => kids.forEach((k) => (k.state = rand() < 0.5 ? STATES.SKIPPED : STATES.UPDATED)),
    // Mixed problem: should stay expanded
    (kids) =>
      kids.forEach(
        (k) =>
          (k.state = pickFrom([STATES.ERROR, STATES.WARNING, STATES.DELAYED, STATES.UPDATING])),
      ),
    // DISABLED + collapsible mix (DISABLED is filtered, so should still collapse via Rule 1)
    (kids) => kids.forEach((k) => (k.state = rand() < 0.4 ? STATES.DISABLED : STATES.UPDATED)),
  ];

  // Apply each pattern to a handful of containers so every rule has
  // multiple samples in the resulting fixture.
  const perPattern = 6;
  let cursor = 0;
  for (const apply of patterns) {
    for (let i = 0; i < perPattern && cursor < leafOnlyContainers.length; i++, cursor++) {
      apply(leafOnlyContainers[cursor].node.children);
    }
  }
}

function summarize(root) {
  const counts = {};
  let leaves = 0;
  for (const leaf of walkLeaves(root)) {
    leaves++;
    counts[leaf.state] = (counts[leaf.state] || 0) + 1;
  }

  // Audit container coverage: how many containers will hit each rule?
  let rule1Collapse = 0; // all same collapsible
  let rule1Stay = 0; // all same non-collapsible
  let rule2 = 0; // only SKIPPED/UPDATED mix
  let mixedStay = 0; // mixed, would stay expanded
  const COLLAPSIBLE = new Set([STATES.READY, STATES.UPDATED, STATES.SKIPPED, STATES.DISABLED]);

  for (const { node } of walkContainers(root)) {
    const childStates = node.children
      .filter((c) => !c.children || !c.children.length) // immediate-leaf check; same shape statusManager uses
      .map((c) => c.state);
    if (!childStates.length) continue;
    const nonDisabled = childStates.filter((s) => s !== STATES.DISABLED);
    if (!nonDisabled.length) continue;
    const unique = [...new Set(nonDisabled)];
    if (unique.length === 1) {
      if (COLLAPSIBLE.has(unique[0])) rule1Collapse++;
      else rule1Stay++;
    } else if (unique.every((s) => s === STATES.SKIPPED || s === STATES.UPDATED)) {
      rule2++;
    } else {
      mixedStay++;
    }
  }

  return { leaves, counts, rule1Collapse, rule1Stay, rule2, mixedStay };
}

async function main() {
  const raw = await fs.readFile(INPUT, 'utf8');

  // No-collapse variation
  const v1 = JSON.parse(raw);
  v1.metadata = {
    ...v1.metadata,
    name: 'All-randomState',
    description: 'Random per-node states for perf testing; auto-collapse disabled.',
    updated: new Date().toISOString(),
  };
  v1.settings = {
    ...v1.settings,
    toggleCollapseOnStatusChange: false,
    cascadeOnStatusChange: false,
  };
  for (const root of v1.nodes) assignRandomStates(root);
  await fs.writeFile(OUT_NO_COLLAPSE, JSON.stringify(v1));
  console.log('Wrote', OUT_NO_COLLAPSE);
  for (const root of v1.nodes) console.log('  summary:', summarize(root));

  // Auto-collapse variation — same shape, same seed, just settings differ
  const v2 = JSON.parse(raw);
  v2.metadata = {
    ...v2.metadata,
    name: 'All-randomState-autoCollapse',
    description: 'Random per-node states for perf testing; auto-collapse + cascade enabled.',
    updated: new Date().toISOString(),
  };
  v2.settings = {
    ...v2.settings,
    toggleCollapseOnStatusChange: true,
    cascadeOnStatusChange: true,
  };
  // Re-seed and re-run so v2 has the same node states as v1 (deterministic).
  // We do this by reimporting the seeded generator side-effects: easier path
  // is to copy v1's node tree wholesale.
  v2.nodes = JSON.parse(JSON.stringify(v1.nodes));
  await fs.writeFile(OUT_COLLAPSE, JSON.stringify(v2));
  console.log('Wrote', OUT_COLLAPSE);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/generate-random-state-variations.py
#!/usr/bin/env python3
"""Generate two perf-test variations of dashboard/data/All.json with random
node states, designed to exercise every auto-collapse rule.

  All-randomState.json              — toggleCollapseOnStatusChange: false
  All-randomState-autoCollapse.json — toggleCollapseOnStatusChange: true

State assignment is seeded so output is reproducible. Distribution and
targeted subtree patches guarantee coverage of:
  - Rule 1 (all children share one collapsible status: READY/UPDATED/SKIPPED/DISABLED)
  - Rule 2 (children are only SKIPPED and/or UPDATED)
  - Mixed-problem containers (ERROR/WARNING/DELAYED) that must stay expanded
"""

import copy
import json
import random
from datetime import datetime, timezone
from pathlib import Path

DATA_DIR = Path(__file__).resolve().parent.parent / "dashboard" / "data"
INPUT = DATA_DIR / "All.json"
OUT_NO_COLLAPSE = DATA_DIR / "All-randomState.json"
OUT_COLLAPSE = DATA_DIR / "All-randomState-autoCollapse.json"

# Status values match NodeStatus in dashboard/js/nodeBase.js
UNDETERMINED = "Undetermined"
UNKNOWN = "Unknown"
DISABLED = "Disabled"
READY = "Ready"
UPDATING = "Updating"
UPDATED = "Updated"
SKIPPED = "Skipped"
DELAYED = "Delayed"
WARNING = "Warning"
ERROR = "Error"

WEIGHTED = [
    (UPDATED, 24),
    (READY, 18),
    (SKIPPED, 14),
    (UPDATING, 10),
    (DELAYED, 8),
    (WARNING, 8),
    (ERROR, 7),
    (DISABLED, 5),
    (UNKNOWN, 3),
    (UNDETERMINED, 3),
]
COLLAPSIBLE = {READY, UPDATED, SKIPPED, DISABLED}


def walk_containers(node):
    if node.get("children"):
        yield node
        for c in node["children"]:
            yield from walk_containers(c)


def walk_leaves(node):
    if node.get("children"):
        for c in node["children"]:
            yield from walk_leaves(c)
    else:
        yield node


def assign_random_states(root, rng):
    states, weights = zip(*WEIGHTED)

    for leaf in walk_leaves(root):
        leaf["state"] = rng.choices(states, weights=weights, k=1)[0]

    leaf_only = [
        n for n in walk_containers(root)
        if n["children"] and all(not c.get("children") for c in n["children"])
    ]
    rng.shuffle(leaf_only)

    def all_to(kids, value):
        for k in kids:
            k["state"] = value

    def mix_skipped_updated(kids):
        for k in kids:
            k["state"] = SKIPPED if rng.random() < 0.5 else UPDATED

    def mix_problem(kids):
        for k in kids:
            k["state"] = rng.choice([ERROR, WARNING, DELAYED, UPDATING])

    def mix_disabled_updated(kids):
        for k in kids:
            k["state"] = DISABLED if rng.random() < 0.4 else UPDATED

    patterns = [
        # Rule 1: all-same collapsible status
        lambda k: all_to(k, READY),
        lambda k: all_to(k, UPDATED),
        lambda k: all_to(k, SKIPPED),
        lambda k: all_to(k, DISABLED),
        # Rule 1: all-same non-collapsible status (must NOT collapse)
        lambda k: all_to(k, ERROR),
        lambda k: all_to(k, WARNING),
        # Rule 2: SKIPPED + UPDATED only mix
        mix_skipped_updated,
        # Mixed problem: should stay expanded
        mix_problem,
        # DISABLED + UPDATED mix (DISABLED is filtered, so should still collapse via Rule 1)
        mix_disabled_updated,
    ]

    per_pattern = 6
    cursor = 0
    for apply in patterns:
        for _ in range(per_pattern):
            if cursor >= len(leaf_only):
                break
            apply(leaf_only[cursor]["children"])
            cursor += 1


def summarize(root):
    counts = {}
    leaves = 0
    for leaf in walk_leaves(root):
        leaves += 1
        counts[leaf["state"]] = counts.get(leaf["state"], 0) + 1

    rule1_collapse = rule1_stay = rule2 = mixed_stay = 0
    for n in walk_containers(root):
        child_states = [c["state"] for c in n["children"] if not c.get("children")]
        if not child_states:
            continue
        non_disabled = [s for s in child_states if s != DISABLED]
        if not non_disabled:
            continue
        unique = set(non_disabled)
        if len(unique) == 1:
            if next(iter(unique)) in COLLAPSIBLE:
                rule1_collapse += 1
            else:
                rule1_stay += 1
        elif unique <= {SKIPPED, UPDATED}:
            rule2 += 1
        else:
            mixed_stay += 1

    return {
        "leaves": leaves,
        "counts": counts,
        "rule1_collapse_candidates": rule1_collapse,
        "rule1_stay_expanded": rule1_stay,
        "rule2_collapse_candidates": rule2,
        "mixed_stay_expanded": mixed_stay,
    }


def main():
    raw = json.loads(INPUT.read_text())
    rng = random.Random(0xFD05_2026)

    v1 = copy.deepcopy(raw)
    v1["metadata"] = {
        **v1.get("metadata", {}),
        "name": "All-randomState",
        "description": "Random per-node states for perf testing; auto-collapse disabled.",
        "updated": datetime.now(timezone.utc).isoformat(),
    }
    v1["settings"] = {
        **v1.get("settings", {}),
        "toggleCollapseOnStatusChange": False,
        "cascadeOnStatusChange": False,
    }
    for root in v1["nodes"]:
        assign_random_states(root, rng)

    OUT_NO_COLLAPSE.write_text(json.dumps(v1, separators=(",", ":")))
    print(f"Wrote {OUT_NO_COLLAPSE}")
    for root in v1["nodes"]:
        print("  summary:", summarize(root))

    v2 = copy.deepcopy(raw)
    v2["metadata"] = {
        **v2.get("metadata", {}),
        "name": "All-randomState-autoCollapse",
        "description": "Random per-node states for perf testing; auto-collapse + cascade enabled.",
        "updated": datetime.now(timezone.utc).isoformat(),
    }
    v2["settings"] = {
        **v2.get("settings", {}),
        "toggleCollapseOnStatusChange": True,
        "cascadeOnStatusChange": True,
    }
    v2["nodes"] = copy.deepcopy(v1["nodes"])
    OUT_COLLAPSE.write_text(json.dumps(v2, separators=(",", ":")))
    print(f"Wrote {OUT_COLLAPSE}")


if __name__ == "__main__":
    main()
